### DISCUSSION TITLE
How to copy an existing website in pb_public to a new website?

### DISCUSSION THREAD

**Discussion by Mashiane** [pocketbase/pocketbase#6190] [Q&A] [answered]
Hi

I want to make available the same web application to different users but each one should have their own working space. This is due to them having their own routes per website. The route creation process is well, thank you again.

I am using pockethost.io for this.

I have deployed a website there which will act as a template website for the multiple user facing websites.

I need to copy the existing website folder in the pb_public folder and then use a new name for it. This will be done via a routeAdd.

`
routerAdd("POST", "/api/createwebsite", (c) => {
    
try {

        //get the data from the request

        const data = $apis.requestInfo(c).data

        //get the account

        const account = data.account

       //define new name

       const newWebsite = "client" + account

       // copy existing website on pb_public to the newWebsite name recursively.

        return c.json(200, { "createwebsite": "Success"})
    } catch (err) {
        return c.json(400, { "createwebsite": "Failure", "error": JSON.stringify(err) })
    }
}, $apis.activityLogger($app))
`

**Comment by thangngoc89** [comment]
Since the JSVM $os namespace doesn’t provide a copy function, you could do this by:
1. Implement readDir recursively based on readDir
2. For each of the file, use readFile and writeFile to copy the file (very inefficient since it read the whole thing to memory)
3. For each of the dir, use mkdir


Or you could use golang hooks to achieve the same thing much faster and more efficient. I learned golang for working with pocketbase coming from a JS background.

**Reply by Mashiane** [reply]
Thanks for the suggestions. Appreciated. I was going through the documentation unfortunately I cannot even get the first step right.

i.e. find the actual path for the pb_public directory. Perhaps its me, I cant seem to find relevant examples of doing this as the documentation just provides class and method names, so its difficult to make sense of some things.

If you happen to have an example it could greatly help. I don't have any knowledge of golang and this is just one thing I need at the moment without having to learn everything about it as I don't plan anything extra at the moment.

**Reply by Mashiane** [reply]
I also tried to see if any of the AI tools could provide green lights, thing is what is on the docs and what they spill out is mhe... so jah.

**Reply by thangngoc89** [reply]
@Mashiane I totally understand. I bite the bullet and learn golang because JS hooks’s developer experience is subpar comparing to golang hook. Because every exposed function is generated from a golang function, you get way better type hints there. 

In any case, you need to look at $os documentation here https://pocketbase.io/jsvm/modules/_os.html

For finding pb_public path, you could try relative path from __hooks


const hooksDir = __hooks

**Reply by Mashiane** [reply]
I am thinking, what if I could just zip the website folder I want to copy and then unzip it to pb_public with a new name via routeAdd code? Is that something that could be doable?

**Reply by thangngoc89** [reply]
@Mashiane zipping could work too but how to unzip is another problem. You need to investigate unzipping in the JS hooks.

**Reply by ganigeorgiev** [reply, marked as answer]
@Mashiane If you are trying to copy a local directory with the JS `pb_hooks` then probably the easiest solution would be to invoke a shell command like `cp` or similar, for example:

```js
// https://pocketbase.io/jsvm/functions/_os.cmd.html
$os.cmd('cp', '-r', src, dst).run()
```

Similar to my warning in your other recent question (https://github.com/pocketbase/pocketbase/discussions/6177), you need to be extremely careful who and what you allow to execute because again PocketBase has no sandbox features out of the box.

In any case, at least for me, your description is confusing and it is not clear what actually is the end goal here (aka. what does _"copy an existing website in pb_public to a new wesite"_ even means?). Keep in mind that no one here have any context what you are doing and you'll have better chance someone to assist you if you elaborate more on your use case and the overall bigger picture.

**Reply by Mashiane** [reply]
Thank you so much @ganigeorgiev and @thangngoc89 , the issue has been solved. You just made my day.

![image](https://github.com/user-attachments/assets/c46ccda7-9091-41fe-8763-0360f797df0e)

Yes, the creation of the folders and the hooks will be for my attention alone and no one elses. This is a SAAS app with each user having their own collections and their own hooks and this final step was for them to have their own website. Each website will use hooks specific to the user and collections specific to the user.

So a user will just go to their own app with something like https://oneplace.pockethost.io/clientdymtklvf5t/index.html

Internally the app will use their own pocketbase collections and thus hooks that speak to those collections. For example.

![image](https://github.com/user-attachments/assets/3df07d41-54b3-425e-b450-b297500853a9)

Again, thank you so much.

**Comment by Mashiane** [comment]
Just one last thing. How can I update the code to ensure that it executes and completes and then returns the json result..

`
routerAdd("POST", "/api/createwebsite", (c) => {
    
try {

        //get the data from the request

        const data = $apis.requestInfo(c).data

        //get the account

        const account = data.account

        //define the pb_public folder

        const src = '/home/pockethost/pb_public/client'

        const dst = '/home/pockethost/pb_public/client' + account

        const src1 = '/home/pockethost/pb_public/server'

        const dst1 = '/home/pockethost/pb_public/server' + account

        //copy the folder

        // https://pocketbase.io/jsvm/functions/_os.cmd.html

        $os.cmd('cp', '-rf', src, dst).run()

        $os.cmd('cp', '-rf', src1, dst1).run()

        return c.json(200, { "createwebsite": "Success"})
    } catch (err) {

        return c.json(400, { "createwebsite": "Failure", "error": JSON.stringify(err) })
    }
}, $apis.activityLogger($app))
`

**Reply by ganigeorgiev** [reply]
I don't understand. `$os.cmd(...).run()` starts the specified command and waits for it to complete.
